In [1]:
import numpy as np
import pandas as pd

In [3]:
movies =pd.read_csv('tmdb_5000_movies.csv')
credits= pd.read_csv('tmdb_5000_credits.csv')

In [4]:
movies.head(1)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                      homepage     id  \
0  http://www.avatarmovie.com/  19995   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   

  original_title                                           overview  \
0         Avatar  In the 22nd century, a paraplegic Marine is di...   

   popularity                               production_companies  \
0  150.437577  [{"name": "Ingenious Film Partners", "id": 289...   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   

                       tagline   title  vote_average  vote_count  
0  Enter the World of Pandora.  Avatar           7.2       11800

In [5]:
credits.head(1)

movie_id   title                                               cast  \
0     19995  Avatar  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [6]:
movies.shape

(4803, 20)

In [9]:
movies.merge(credits,on='title') # here i am merging the two dataframes on the title column. This will give me a new dataframe that has all the columns from both dataframes. I can then use this new dataframe to do my analysis.

budget                                             genres  \
0     237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1     300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2     245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3     250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4     260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
...         ...                                                ...   
4816     220000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4817       9000  [{"id": 35, "name": "Comedy"}, {"id": 10749, "...   
4818          0  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...   
4819          0                                                 []   
4820          0                [{"id": 99, "name": "Documentary"}]   

                                               homepage      id  \
0                           http://www.avatarmovie.com/   19995   
1          http://disney.go.com/disneypictures/pirates/     285   
2           http://www.sonypictures.com/movies/spectre/  206647   
3                    http://www.thedarkknightrises.com/   49026   
4                  http://movies.disney.com/john-carter   49529   
...                                                 ...     ...   
4816                                                NaN    9367   
4817                                                NaN   72766   
4818  http://www.hallmarkchannel.com/signedsealeddel...  231617   
4819                        http://shanghaicalling.com/  126186   
4820                                                NaN   25975   

                                               keywords original_language  \
0     [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1     [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2     [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3     [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4     [{"id": 818, "name": "based on novel"}, {"id":...                en   
...                                                 ...               ...   
4816  [{"id": 5616, "name": "united states\u2013mexi...                es   
4817                                                 []                en   
4818  [{"id": 248, "name": "date"}, {"id": 699, "nam...                en   
4819                                                 []                en   
4820  [{"id": 1523, "name": "obsession"}, {"id": 224...                en   

                                original_title  \
0                                       Avatar   
1     Pirates of the Caribbean: At World's End   
2                                      Spectre   
3                        The Dark Knight Rises   
4                                  John Carter   
...                                        ...   
4816                               El Mariachi   
4817                                 Newlyweds   
4818                 Signed, Sealed, Delivered   
4819                          Shanghai Calling   
4820                         My Date with Drew   

                                               overview  popularity  \
0     In the 22nd century, a paraplegic Marine is di...  150.437577   
1     Captain Barbossa, long believed to be dead, ha...  139.082615   
2     A cryptic message from Bond’s past sends him o...  107.376788   
3     Following the death of District Attorney Harve...  112.312950   
4     John Carter is a war-weary, former military ca...   43.926995   
...                                                 ...         ...   
4816  El Mariachi just wants to play his guitar and ...   14.269792   
4817  A newlywed couple's honeymoon is upended by th...    0.642552   
4818  "Signed, Sealed, Delivered" introduces a dedic...    1.444476   
4819  When ambitious New York attorney Sam is sent t...    0.857008   
4820  Ever since the second grade when he first saw ... 

In [ ]:
movies=movies.merge(credits,on='title') # make sure to save the merged dataframe back to movies

In [10]:
movies.head(5)

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  ... runtime  \
0  [{"name": "Ingenious Film Partners", "id": 289...  ...   162.0   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...  ...   169.0   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...  ...   148.0   
3  [{"name": "Legendary Pictures", "id": 923}, {"...  ...   165.0   
4        [{"name": "Walt Disney Pictures", "id": 2}]  ...   132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title vote_average vote_count movie_id  \
0                                    Avatar          7.2      11800    19995   
1  Pirates of the Caribbean: At World's End          6.9       4500      285   
2                                   Spectre          6.3       4466   206647   
3                     The Dark Knight Rises          7.6       9106    49026   
4                               John Carter          6.1       2124    49529   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...

In [9]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4809 entries, 0 to 4808
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   budget                4809 non-null   int64  
 1   genres                4809 non-null   object 
 2   homepage              1713 non-null   object 
 3   id                    4809 non-null   int64  
 4   keywords              4809 non-null   object 
 5   original_language     4809 non-null   object 
 6   original_title        4809 non-null   object 
 7   overview              4806 non-null   object 
 8   popularity            4809 non-null   float64
 9   production_companies  4809 non-null   object 
 10  production_countries  4809 non-null   object 
 11  release_date          4808 non-null   object 
 12  revenue               4809 non-null   int64  
 13  runtime               4807 non-null   float64
 14  spoken_languages      4809 non-null   object 
 15  status               

In [ ]:
movies = movies[['movie_id','title','overview','genres','keywords','cast','crew']] # here i am selecting only the columns that i need for my analysis. I am dropping the other columns that are not relevant for my analysis.

In [11]:
movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                              genres  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [12]:
movies.isnull().sum() # here i am checking for null values in the dataframe. This will give me an idea of how many null values are there in each column. I can then decide how to handle these null values. For example, if there are a lot of null values in a column, I might want to drop that column. If there are only a few null values, I might want to fill them with some value (like the mean or median).

movie_id    0
title       0
overview    3
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [ ]:
movies.dropna(inplace=True) #make sure to drop the null values from the dataframe. This will give me a clean dataframe that I can use for my analysis.

In [ ]:
movies.isnull().sum() 

movie_id    0
title       0
overview    0
genres      0
keywords    0
cast        0
crew        0
dtype: int64

In [ ]:
import ast  # here i am importing the ast module. This module will help me to convert the string representation of a list into an actual list. For example, if i have a string like this: "[1, 2, 3]", i can use ast.literal_eval() to convert it into a list like this: [1, 2, 3].

In [ ]:
def convert(text): # this function will take a string representation of a list and convert it into an actual list. For example, if i have a string like this: "[1, 2, 3]", this function will convert it into a list like this: [1, 2, 3].
    L=[]
    for i in ast.literal_eval(text):
        L.append(i['name'])   # here i am appending the name of the genre, keyword, cast member, or crew member to the list L. This is because i only want to consider the name of the genre, keyword, cast member, or crew member for my analysis. I don't care about the other information that is stored in the string (like the id or the popularity).
    return L
#string repre and actual list means that sometimes when we read data from a csv file, the list is stored as a string. For example, if we have a list like this: [1, 2, 3], it might be stored in the csv file as a string like this: "[1, 2, 3]". This function will take that string and convert it back into an actual list that we can use for our analysis.

In [ ]:
movies['genres']=movies['genres'].apply(convert) # here i am applying the convert function to the genres column. This will convert the string representation of the list into an actual list that we can use for our analysis.
movies.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                            overview  \
0  In the 22nd century, a paraplegic Marine is di...   
1  Captain Barbossa, long believed to be dead, ha...   
2  A cryptic message from Bond’s past sends him o...   
3  Following the death of District Attorney Harve...   
4  John Carter is a war-weary, former military ca...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   
1                   [Adventure, Fantasy, Action]   
2                     [Action, Adventure, Crime]   
3               [Action, Crime, Drama, Thriller]   
4           [Action, Adventure, Science Fiction]   

                                            keywords  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4  [{"id": 818, "name": "based on novel"}, {"id":...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [ ]:
movies['keywords'] = movies['keywords'].apply(convert) # here i am applying the convert function to the keywords column. This will convert the string representation of the list into an actual list that we can use for our analysis.
movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [ ]:
def convert3(text):  # only give me the top 3 elements from the list. This is because some movies have a lot of cast members and crew members, and i only want to consider the top 3 for my analysis.
    L=[]
    count=0
    for i in ast.literal_eval(text):        
        if count < 3:
            L.append(i['name'])
            count+=1
    return L

In [ ]:
movies['cast'] = movies['cast'].apply(convert3)  # here i am applying the convert3 function to the cast column. This will convert the string representation of the list into an actual list that we can use for our analysis. It will also only give me the top 3 cast members for each movie.

In [21]:
movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                               cast  \
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...

In [22]:
def find_director(text):
    L=[]
    for i in ast.literal_eval(text):
        if i['job'] == 'Director':
            L.append(i['name'])
            break
    return L

In [23]:
movies['crew']=movies['crew'].apply(find_director)
movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  In the 22nd century, a paraplegic Marine is di...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                               cast             crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]  [James Cameron]

In [ ]:
movies['overview'][0]  # index 0 is the first movie in the dataframe. This will give me the overview of the first movie in the dataframe.
# this is in string format. I want to convert it into a list of words. This is because i want to use the overview for my analysis, and it will be easier to work with a list of words than a string. I can use the split() method to convert the string into a list of words. For example, if i have a string like this: "This is an overview of the movie.", i can use the split() method to convert it into a list like this: ["This", "is", "an", "overview", "of", "the", "movie."].

'In the 22nd century, a paraplegic Marine is dispatched to the moon Pandora on a unique mission, but becomes torn between following orders and protecting an alien civilization.'

string converted to list 
also i used lambda function here because i want to apply the split() method to each element in the overview column. The lambda function allows me to do that in a concise way.
here i am applying the split() method to the overview column. This will convert the string representation of the overview into a list of words that we can use for our analysis.

In [ ]:
movies['overview'] = movies['overview'].apply(lambda x:x.split())   # here i am applying the split() method to the overview column. This will convert the string representation of the overview into a list of words that we can use for our analysis.

In [ ]:
movies.head(1) # now see all the columns have been converted into the desired format. The genres, keywords, cast, and crew columns are now lists of names, and the overview column is now a list of words. This will make it easier for me to do my analysis and build my movie recommender system.

movie_id   title                                           overview  \
0     19995  Avatar  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                          genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  \
0  [culture clash, future, space war, space colon...   

                                               cast             crew  
0  [Sam Worthington, Zoe Saldana, Sigourney Weaver]  [James Cameron]

let remove space from the names in the genres, keywords, cast, and crew columns. This is because when we will create a new column called 'tags' by combining all these columns, we want to make sure that there are no spaces in the names. This will help us to create a better tags column that we can use for our analysis. 


In [ ]:
# let remove space from the names in the genres, keywords, cast, and crew columns. This is because when we will create a new column called 'tags' by combining all these columns, we want to make sure that there are no spaces in the names. This will help us to create a better tags column that we can use for our analysis. 
movies['genres'] = movies['genres'].apply(lambda x:[i.replace(" ","") for i in x])
movies['keywords'] = movies['keywords'].apply(lambda x:[i.replace(" ","") for i in x])
movies['cast'] = movies['cast'].apply(lambda x:[i.replace(" ","") for i in x])
movies['crew'] = movies['crew'].apply(lambda x:[i.replace(" ","") for i in x])


In [28]:
movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   

                                            cast            crew  
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]  [JamesCameron]

In [ ]:
movies['tags'] = movies['overview'] + movies['genres'] + movies['keywords'] + movies['cast']+ movies['crew'] # here i am creating a new column called 'tags' by combining the overview, genres, keywords, cast, and crew columns. This will give me a single column that contains all the information that i need for my analysis. I can then use this tags column to build my movie recommender system.

In [30]:
movies.head(1)

movie_id   title                                           overview  \
0     19995  Avatar  [In, the, 22nd, century,, a, paraplegic, Marin...   

                                         genres  \
0  [Action, Adventure, Fantasy, ScienceFiction]   

                                            keywords  \
0  [cultureclash, future, spacewar, spacecolony, ...   

                                            cast            crew  \
0  [SamWorthington, ZoeSaldana, SigourneyWeaver]  [JamesCameron]   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...

In [31]:
new_df=movies[['movie_id','title','tags']]

In [32]:
new_df.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                tags  
0  [In, the, 22nd, century,, a, paraplegic, Marin...  
1  [Captain, Barbossa,, long, believed, to, be, d...  
2  [A, cryptic, message, from, Bond’s, past, send...  
3  [Following, the, death, of, District, Attorney...  
4  [John, Carter, is, a, war-weary,, former, mili...

In [ ]:
 new_df['tags']= new_df['tags'].apply(lambda x:" ".join(x))   # here i am applying the join() method to the tags column. This will convert the list of words in the tags column into a single string that we can use for our analysis. For example, if we have a list like this: ["This", "is", "a", "tag"], it will be converted into a string like this: "This is a tag". This will make it easier for us to use the tags for our analysis and build our movie recommender system.
# also i used lambda function here because i want to apply the join() method to each element in the tags column. The lambda function allows me to do that in a concise way.

C:\Users\ranje\AppData\Local\Temp\ipykernel_14240\507474843.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags']= new_df['tags'].apply(lambda x:" ".join(x))   #converted to string


In [34]:
new_df.head(2)

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   

                                                tags  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...

In [ ]:
import nltk  # here i am importing the nltk library. This library will help me to do natural language processing tasks like stemming and lemmatization. Stemming is the process of reducing a word to its root form. For example, the word "running" will be reduced to "run". Lemmatization is the process of reducing a word to its base form. For example, the word "running" will be reduced to "run". I will use stemming in this project to reduce the words in the tags column to their root form. This will help me to create a better tags column that we can use for our analysis.

In [ ]:
from nltk.stem.porter import PorterStemmer  # here i am importing the PorterStemmer class from the nltk.stem.porter module. This class will help me to do stemming on the words in the tags column. Stemming is the process of reducing a word to its root form. For example, the word "running" will be reduced to "run". I will use stemming in this project to reduce the words in the tags column to their root form. This will help me to create a better tags column that we can use for our analysis.
ps = PorterStemmer()

In [ ]:
def stem(text): # i created this function to do stemming on the words in the tags column. This function will take a string as input, split it into words, stem each word using the PorterStemmer, and then join the stemmed words back into a single string. This will help me to create a better tags column that we can use for our analysis.
    y=[]
    for i in text.split():
        y.append(ps.stem(i))
    return " ".join(y)

In [ ]:
new_df['tags']=new_df['tags'].apply(stem)   # here i am applying the stem function to the tags column. This will stem each word in the tags column and then join the stemmed words back into a single string. This will help me to create a better tags column that we can use for our analysis.

C:\Users\ranje\AppData\Local\Temp\ipykernel_14240\3514595201.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_df['tags']=new_df['tags'].apply(stem)


In [ ]:
from sklearn.feature_extraction.text import CountVectorizer   #countvectorizer is a class in the sklearn library that will help me to convert the text data in the tags column into a matrix of token counts. This will allow me to use the tags for my analysis and build my movie recommender system.
cv = CountVectorizer(max_features = 5000, stop_words='english') # created a object cvd --- IGNORE ---

In [ ]:
vector = cv.fit_transform(new_df['tags']).toarray()  #numpy array mei convert kiya #applying fit_transform() method to the tags column. This will convert the text data in the tags column into a matrix of token counts. The toarray() method will convert the sparse matrix into a numpy array that we can use for our analysis.

now we have a matrix of token counts for each movie in the tags column. Each row in the matrix represents a movie, and each column represents a word in the tags. The value in each cell represents the number of times that word appears in the tags for that movie. For example, if the value in the cell is 3, it means that the word appears 3 times in the tags for that movie. If the value is 0, it means that the word does not appear in the tags for that movie.


In [ ]:
vector[0] # there will many zeros(Sparsity )as only few words available for one movies # in short this is the vector representation of the tags column for the first movie in the dataframe. Each element in the vector represents the count of a particular word in the tags column for that movie. Since there are many words in the tags column, there will be many zeros in the vector because most movies will not have all the words in their tags. This is known as sparsity, and it is a common issue when working with text data.

array([0, 0, 0, ..., 0, 0, 0])

In [ ]:
cv.get_feature_names_out()  # this will give me the list of all the words that are present in the tags column. This is the list of features that we will use for our analysis. Each word in this list corresponds to a column in the vector representation of the tags column. For example, if the word "action" is present in this list, then there will be a column in the vector representation that represents the count of the word "action" in the tags column for each movie.

array(['000', '007', '10', ..., 'zone', 'zoo', 'zooeydeschanel'],
      dtype=object)

In [43]:
vector.shape

(4806, 5000)

there is 4806 movies/vectors and 5000 words taken so, for recommendations we choose closes vectors/movies. vectors basically stores the  data of counts of words of tags.......
2. we will use cosine distance and Euclidean(Due to high Dimenstions)

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity  # here i am importing the cosine_similarity function from the sklearn.metrics.pairwise module. This function will help me to calculate the cosine similarity between the vector representations of the tags column for each movie. The cosine similarity is a measure of similarity between two vectors. It ranges from -1 to 1, where 1 means that the vectors are identical, 0 means that the vectors are orthogonal (i.e., they have no similarity), and -1 means that the vectors are opposite (i.e., they have maximum dissimilarity). I will use the cosine similarity to find similar movies based on their tags.


In [ ]:
similarity = cosine_similarity(vector)  #calcu similarity between the vector representations of the tags column for each movie. This will give me a matrix of cosine similarity values, where each element (i, j) in the matrix represents the cosine similarity between the tags of movie i and movie j. I can then use this matrix to find similar movies based on their tags. For example, if I want to find movies that are similar to a particular movie, I can look at the corresponding row in the similarity matrix and find the movies that have high cosine similarity values with that movie.


In [ ]:
similarity[0]   # distance b/w first movie with each movie, diagonal will always be one , similarity[0][0] # distance b/w first movie with itself, it will always be one because the cosine similarity of a vector with itself is always 1. This is because the angle between the vector and itself is 0 degrees, and the cosine of 0 degrees is 1. Therefore, the diagonal elements of the similarity matrix will always be 1, indicating that each movie is perfectly similar to itself.

array([1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,
       0.        ])

In [ ]:
similarity # this will give me the entire similarity matrix. Each element (i, j) in the matrix represents the cosine similarity between the tags of movie i and movie j. I can use this matrix to find similar movies based on their tags. For example, if I want to find movies that are similar to a particular movie, I can look at the corresponding row in the similarity matrix and find the movies that have high cosine similarity values with that movie.

array([[1.        , 0.08346223, 0.0860309 , ..., 0.04499213, 0.        ,
        0.        ],
       [0.08346223, 1.        , 0.06063391, ..., 0.02378257, 0.        ,
        0.02615329],
       [0.0860309 , 0.06063391, 1.        , ..., 0.02451452, 0.        ,
        0.        ],
       ...,
       [0.04499213, 0.02378257, 0.02451452, ..., 1.        , 0.03962144,
        0.04229549],
       [0.        , 0.        , 0.        , ..., 0.03962144, 1.        ,
        0.08714204],
       [0.        , 0.02615329, 0.        , ..., 0.04229549, 0.08714204,
        1.        ]])

In [ ]:
# list(enumerate(similarity[6])) makes sure indexing
#key=lambda x:x[1] means that we want to sort the list based on the second element of the tuple, which is the cosine similarity value. This will help us to sort the movies based on their similarity to the 7th movie.
# reverse = True means that we want to sort the list in descending order based on the cosine similarity values. This will help us to get the most similar movies at the top of the list.
sorted(list(enumerate(similarity[6 ])), reverse = True,key=lambda x:x[1])[1:6] #  sorting the list of cosine similarity values for the 7th movie in descending order. The enumerate() function is used to get the index of each movie along with its cosine similarity value. The sorted() function is used to sort the list based on the cosine similarity values (which is the second element of the tuple). The reverse=True argument is used to sort in descending order. Finally, [1:6] is used to get the top 5 similar movies (excluding the first one which is the movie itself). This will give me a list of tuples, where each tuple contains the index of a similar movie and its cosine similarity value with the 7th movie.

[(31, np.float64(0.3547874375934496)),
 (79, np.float64(0.3474411464169823)),
 (68, np.float64(0.33593550657351257)),
 (129, np.float64(0.32304786919046097)),
 (16, np.float64(0.317925392228937))]

In [ ]:
# let make funtion for recommendation so that we can easily get the similar movies by just passing the movie name as an argument to the function. This will make it easier for us to use the movie recommender system and get similar movies based on the tags.
def recommend(movie):  # this function will take a movie title as input and recommend 5 similar movies based on the cosine similarity of the tags. The function will first find the index of the movie in the new_df dataframe, then it will get the corresponding row from the similarity matrix, sort the values in descending order, and finally return the titles of the top 5 similar movies (excluding the first one which is the movie itself).
    index = new_df[new_df['title'] == movie].index[0]
    distances = sorted(list(enumerate(similarity[index])),reverse=True,key = lambda x: x[1])
    for i in distances[1:6]:
        print(new_df.iloc[i[0]].title)

In [137]:
recommend('Avatar')

Aliens vs Predator: Requiem
Aliens
Falcon Rising
Independence Day
Titan A.E.


In [58]:
new_df

movie_id                                     title  \
0        19995                                    Avatar   
1          285  Pirates of the Caribbean: At World's End   
2       206647                                   Spectre   
3        49026                     The Dark Knight Rises   
4        49529                               John Carter   
...        ...                                       ...   
4804      9367                               El Mariachi   
4805     72766                                 Newlyweds   
4806    231617                 Signed, Sealed, Delivered   
4807    126186                          Shanghai Calling   
4808     25975                         My Date with Drew   

                                                   tags  
0     in the 22nd century, a parapleg marin is dispa...  
1     captain barbossa, long believ to be dead, ha c...  
2     a cryptic messag from bond’ past send him on a...  
3     follow the death of district attorney harvey d...  
4     john carter is a war-weary, former militari ca...  
...                                                 ...  
4804  el mariachi just want to play hi guitar and ca...  
4805  a newlyw couple' honeymoon is upend by the arr...  
4806  "signed, sealed, delivered" introduc a dedic q...  
4807  when ambiti new york attorney sam is sent to s...  
4808  ever sinc the second grade when he first saw h...  

[4806 rows x 3 columns]

In [13]:
#Now i have to give to my site

In [ ]:
import pickle   # here i am importing the pickle module. This module will help me to save my model and the data that i have used for my analysis. I will use pickle to save the new_df dataframe and the similarity matrix, so that i can use them later when i will deploy my movie recommender system on a website.

In [ ]:
pickle.dump(new_df, open('movies.pkl','wb')) # here i am using the dump() method of the pickle module to save the new_df dataframe to a file called 'movies.pkl'. The 'wb' argument stands for "write binary", which means that the file will be opened in binary mode for writing. This will allow me to save the new_df dataframe in a format that can be easily loaded later when i will deploy my movie recommender system on a website.

In [ ]:
new_df['title'].values       # this will give me the array of movie titles that are present in the new_df dataframe. This is the list of movie titles that we will use for our analysis and to build our movie recommender system. Each title in this array corresponds to a row in the new_df dataframe, and it will be used to find similar movies based on the tags column.

array(['Avatar', "Pirates of the Caribbean: At World's End", 'Spectre',
       ..., 'Signed, Sealed, Delivered', 'Shanghai Calling',
       'My Date with Drew'], dtype=object)

In [ ]:
pickle.dump(similarity,open('similarity.pkl','wb'))  # here i am using the dump() method of the pickle module to save the similarity matrix to a file called 'similarity.pkl'. The 'wb' argument stands for "write binary", which means that the file will be opened in binary mode for writing. This will allow me to save the similarity matrix in a format that can be easily loaded later when i will deploy my movie recommender system on a website.

In [ ]:
# i made two files movies.pkl and similarity.pkl. The movies.pkl file contains the new_df dataframe, which has the movie_id, title, and tags columns. The similarity.pkl file contains the similarity matrix, which has the cosine similarity values between the tags of each movie. I will use these two files to build my movie recommender system on a website. When a user inputs a movie title, I will load the new_df dataframe and the similarity matrix, find the index of the input movie in the new_df dataframe, and then use the similarity matrix to find similar movies based on their tags. Finally, I will return the titles of the similar movies to the user.